lib/app/sequence_base.rb

```diff
@@ -708,10 +710,17 @@ def validate_reference_resolutions(resource, resolved_references = Set.new, max_
                 next
               end
             end
-            value.read
+            reference = value.reference
+            reference_type = value.resource_type
+            resource = value.read
+
+            raise InvalidReferenceResource if resource.resourceType != reference_type
+
```

Comment:
Is there a unit test for this newly introduced logic?

Reply:
I didn't add any unit tests for this, but I did have to update some existing unit tests which were failing because they were returning incorrect resource types.

Reply:
Is it convenient to add some unit tests for this section?

Reply:
Added.